---
title: "Activity 4.1 - Web Scraping"
author: "Tyler Allen"
date: "2024-10-15"
format:
  html:
    cold-fold: true
    embed-resources: true
theme: "lux"
---

# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [1]:
import pandas as pd

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [2]:
import requests
response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

Use Beautiful Soup to parse this string into a tree called `soup`

In [3]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [11]:
tables = soup.find_all("table")
len(tables)

10

In [13]:
#did this to find a style I could use for the next part
tables = soup.find_all("table")
for i, table in enumerate(tables):
    print(f"Table {i}:")
    print(table.attrs)

Table 0:
{'class': ['sidebar', 'nomobile', 'nowraplinks']}
Table 1:
{'class': ['wikitable']}
Table 2:
{'class': ['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}
Table 3:
{'class': ['wikitable']}
Table 4:
{'class': ['wikitable'], 'style': 'text-align:left'}
Table 5:
{'class': ['sortable', 'wikitable', 'static-row-numbers', 'sort-under', 'col1left'], 'style': 'text-align:right'}
Table 6:
{'class': ['wikitable', 'sortable', 'sort-under']}
Table 7:
{'class': ['wikitable', 'sortable', 'sort-under', 'col1left', 'col2center'], 'style': 'text-align:right'}
Table 8:
{'class': ['nowraplinks', 'mw-collapsible', 'mw-collapsed', 'navbox-inner'], 'style': 'border-spacing:0;background:transparent;color:inherit'}
Table 9:
{'class': ['navbox-columns-table'], 'style': 'border-spacing: 0px; text-align:left;width:auto; margin-left:auto; margin-right:auto;'}


We can use attributes like `class=` and `style=` to narrow down the list.

In [26]:
tables = soup.find_all("table", class_=['sortable', 'wikitable', 'sticky-header-multi', 'static-row-numbers', 'sort-under', 'col1left', 'col2center'], attrs={'style': 'text-align:right'})
len(tables)

3

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [27]:
table = tables[0]
table

<table class="sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center" style="text-align:right">
<tbody><tr>
<th rowspan="2">City
</th>
<th rowspan="2"><abbr title="State or district">ST</abbr>
</th>
<th rowspan="2">2023<br/>estimate
</th>
<th rowspan="2">2020<br/>census
</th>
<th rowspan="2">Change
</th>
<th colspan="2">2020 land area
</th>
<th colspan="2">2020 density
</th>
<th rowspan="2">Location
</th></tr>
<tr class="static-row-header">
<th>mi<sup>2</sup>
</th>
<th>km<sup>2</sup>
</th>
<th>/ mi<sup>2</sup>
</th>
<th>/ km<sup>2</sup>
</th></tr>
<tr>
<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>
<td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
</td>
<td style="text-align:right;">8,258,035
</td>
<td style="text-align:right;

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

In [39]:
#initialize an empty list
rows = []

#iterate over all rows in the table (skipping the header row)
for row in table.find_all("tr")[1:]:
    # Get all the cells (<td>) in the row
    cells = row.find_all("td")

    #check if there are enough cells to extract the needed information so there isn't an error
    if len(cells) >= 4:
        #extract city name from cell[0]
        city = cells[0].text.strip()
        #print(f"City: {city}")

        #extract state from cell[1]
        state = cells[1].text.strip()
        #print(f"State: {state}")

        #extract population (2022 estimate) from cell[2] and remove commas
        population = cells[2].text.strip().replace(',', '')
        #print(f"Population: {population}")

        #extract 2020 land area (sq mi) from cell[3] and remove commas
        land_area = cells[3].text.strip().replace(',', '')
        #print(f"Land Area: {land_area}")

        #append this data
        rows.append({
            "city": city,
            "state": state,
            "population (2022 estimate)": population,
            "2020 land area (sq mi)": land_area
        })
#convert to pandas df
df = pd.DataFrame(rows)

df

city state population (2022 estimate) 2020 land area (sq mi)
0    New York[c]    NY                    8258035                8804190
1    Los Angeles    CA                    3820914                3898747
2        Chicago    IL                    2664452                2746388
3        Houston    TX                    2314157                2304580
4        Phoenix    AZ                    1650070                1608139
..           ...   ...                        ...                    ...
331         Yuma    AZ                     100858                  95548
332  New Bedford    MA                     100695                 101079
333   Suffolk[l]    VA                     100659                  94324
334     Hesperia    CA                     100633                  99818
335    Davenport    IA                     100354                 101724

[336 rows x 4 columns]

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [40]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

[                    Population tables of U.S. cities
 0  The skyline of New York City, the most populou...
 1                                             Cities
 2  Population AreaDensityEthnic identityForeign-b...
 3                                        Urban areas
 4             Populous cities and metropolitan areas
 5                                 Metropolitan areas
 6  184 combined statistical areas935 core-based s...
 7                                        Megaregions
 8  See related population listsNorth American met...
 9                                                vte,
     0                               1
 0 NaN                   State capital
 1 NaN              State largest city
 2 NaN  State capital and largest city
 3 NaN                 Federal capital,
             City  ST 2023 estimate 2020 census  Change 2020 land area          \
             City  ST 2023 estimate 2020 census  Change            mi2     km2   
 0    New York[c]  NY       8258035     88041

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [56]:
#changed text-align to right and took out 'sortable'
tables = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable', "style": "text-align:right"})
len(tables)

3

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [51]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': 'wikitable', "style": "text-align:right"})[0]
df_cities2

City  ST 2023 estimate 2020 census  Change 2020 land area          \
            City  ST 2023 estimate 2020 census  Change            mi2     km2   
0    New York[c]  NY       8258035     8804190  −6.20%          300.5   778.3   
1    Los Angeles  CA       3820914     3898747  −2.00%          469.5  1216.0   
2        Chicago  IL       2664452     2746388  −2.98%          227.7   589.7   
3        Houston  TX       2314157     2304580  +0.42%          640.4  1658.6   
4        Phoenix  AZ       1650070     1608139  +2.61%          518.0  1341.6   
..           ...  ..           ...         ...     ...            ...     ...   
331         Yuma  AZ        100858       95548  +5.56%          120.7   312.6   
332  New Bedford  MA        100695      101079  −0.38%           20.0    51.8   
333   Suffolk[l]  VA        100659       94324  +6.72%          399.2  1033.9   
334     Hesperia  CA        100633       99818  +0.82%           72.7   188.3   
335    Davenport  IA        100354      101724  −1.35%           63.8   165.2   

    2020 density                                      Location  
           / mi2  / km2                               Location  
0          29298  11312    40°40′N 73°56′W / 40.66°N 73.94°W  
1           8304   3206  34°01′N 118°25′W / 34.02°N 118.41°W  
2          12061   4657    41°50′N 87°41′W / 41.84°N 87.68°W  
3           3599   1390    29°47′N 95°23′W / 29.79°N 95.39°W  
4           3105   1199  33°34′N 112°05′W / 33.57°N 112.09°W  
..           ...    ...                                    ...  
331          792    306  32°31′N 114°31′W / 32.52°N 114.52°W  
332         5054   1951    41°40′N 70°56′W / 41.66°N 70.94°W  
333          236     91    36°42′N 76°38′W / 36.70°N 76.63°W  
334         1373    530  34°24′N 117°19′W / 34.40°N 117.32°W  
335         1594    615    41°34′N 90°36′W / 41.56°N 90.60°W  

[336 rows x 10 columns]

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [54]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag

In [63]:
response = requests.get("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

# Parse the HTML content
soup = BeautifulSoup(response.content, "html.parser")

#Find all course blocks (instead of tables)
course_blocks = soup.find_all("div", {"class": "courseblock"})

#Initialize a list to store course data
courses_data = []

#Iterate through each course block to extract the required information
for course_block in course_blocks:
    #Course name and number
    course_title_tag = course_block.find("p", {"class": "courseblocktitle"})
    if course_title_tag:
        course_title = course_title_tag.get_text(strip=True)

        # Remove the units part if it exists
        units_tag = course_title_tag.find("span")
        if units_tag:
            units_tag.extract()
        course_title = course_title_tag.get_text(strip=True)

        # Extract the term typically offered
        term_offered_tag = course_block.find("p", {"class": "noindent"})
        if term_offered_tag:
            term_offered = term_offered_tag.get_text(strip=True)
            #used ChatGPT to help me so it doesn't say "Term Typically Offered" in every row under that column
            term_offered = term_offered.replace("Term Typically Offered: ", "")

            # Append if the course starts with DATA or STAT
            if course_title.startswith("DATA") or course_title.startswith("STAT"):
                courses_data.append({
                    "Course Name and Number": course_title,
                    "Term Typically Offered": term_offered
                })

#Convert into pandas df
df_courses = pd.DataFrame(courses_data)

df_courses

Course Name and Number Term Typically Offered
0                   DATA 100. Data Science for All I.               F, W, SP
1             DATA 301. Introduction to Data Science.               F, W, SP
2          DATA 401. Data Science Process and Ethics.                      F
3   DATA 402. Mathematical Foundations of Data Sci...                      F
4         DATA 403. Data Science Projects Laboratory.                      F
..                                                ...                    ...
69             STAT 551. Statistical Learning with R.                      F
70           STAT 566. Graduate Consulting Practicum.                     SP
71                STAT 570. Selected Advanced Topics.                    TBD
72          STAT 590. Graduate Seminar in Statistics.               F, W, SP
73                                  STAT 599. Thesis.                    TBD

[74 rows x 2 columns]